spec/support/lti_controller_examples.rb

```diff
@@ -0,0 +1,136 @@
+             params: { lti_message_hint: lti_message_hint, target_link_uri: target_link_uri, login_hint: login_hint }
+        is_expected.to respond_with(:unprocessable_entity)
+      end
+      it 'responds with unprocessable_entity if target_link_uri is not passed' do
+        post :launch, params: { lti_message_hint: lti_message_hint, client_id: client_id, login_hint: login_hint }
+        is_expected.to respond_with(:unprocessable_entity)
+      end
+      it 'responds with unprocessable_entity if login_hint is not passed' do
+        post :launch,
+             params: { lti_message_hint: lti_message_hint, client_id: client_id, target_link_uri: target_link_uri }
+        is_expected.to respond_with(:unprocessable_entity)
+      end
+      context 'when all required params exist' do
+        before :each do
+          stub_request(:post, 'http://canvas.instructure.com:443/api/lti/authorize_redirect')
+            .with(
+              body: hash_including({ client_id: 'LMS defined ID',
+                                     login_hint: 'another opque string',
+                                     lti_message_hint: 'opaque string',
+                                     prompt: 'none',
+                                     redirect_uri: 'https://example.com/authorize_redirect',
+                                     response_mode: 'form_post',
+                                     response_type: 'id_token',
+                                     scope: 'openid' }),
+              headers: {
+                Accept: '*/*'
+              }
+            )
+            .to_return(status: 302, body: 'stubbed response', headers: { location: root_url })
+        end
+        context 'with correct parameters' do
+          before { controller.request.headers.merge(HTTP_REFERER: host) }
+          it 'redirects to the host auth url' do
+            post :launch, params: { lti_message_hint: lti_message_hint,
+                                    login_hint: login_hint,
+                                    client_id: client_id, target_link_uri: target_link_uri }
+            expect(response).to redirect_to(root_url)
+          end
+        end
+      end
+    end
+  end
+  describe '#redirect_login' do
+    before :each do
+      controller.request.headers.merge(HTTP_REFERER: host)
+    end
+    context 'with incorrect or missing parameters' do
+      it 'redirects to an error page with no params' do
+        post :redirect_login, params: {}
+        is_expected.to render_template('shared/http_status')
+      end
+      it 'redirects to an error page with a mismatched state' do
+        post :redirect_login, params: { state: state, id_token: 'token' }
+        is_expected.to render_template('shared/http_status')
+      end
+    end
+    context 'with correct parameters' do
+      let(:jwk_url) { 'https://canvas.instructure.com:443/api/lti/security/jwks' }
+      let(:payload) do
+        { aud: client_id,
+          iss: 'https://canvas.instructure.com',
```

Comment:
if this is a shared example for all lti controller subclasses, do not use canvas specific examples